src/background/host.ts

```diff
@@ -34,23 +36,148 @@ export class Host {
+    /**
+     * Subscribe to friend changes
+     * @param cb The callback to call when the friends list changes
+     */
+    public subscribeToFriendChanges(cb: (count: number, delta: number) => void): void {
+        this.friendChangeSubscribers.push(cb);
+    }
+
+    /**
+     * Connects the current host to the given peer
+     * @param hostId The id of the host to connect to
+     */
+    public connectToHost(hostId: string): void {
         const conn = this.#peer.connect(hostId);
 
-        conn.on("open", () => this.#handleConnect(conn));
+        // Handle the conenction to host but don't notify friend changes until the host sends the friend message
+        conn.on("open", () => this.handleConnect(conn, false));
+    }
+
+    /**
+     * Handle a new peer connection
+     * @param conn The underlying data connection
+     */
+    private handleConnect(conn: Peer.DataConnection, shouldNotify = true): void {
+        const friend: Friend = new Friend(conn.peer, undefined, conn);
+        this.#friends.set(friend.id, friend);
+        shouldNotify && this.notifyFriendChanges(1);
+
+        conn.on("data", (data: Message) => {
+            switch (data.messageType) {
+                case MessageType.Friend:
+                    this.handleFriendMessage(data);
+                    break;
+                default:
+                    break;
+            }
+
+            this.#port.postMessage(data);
+        });
+
+        conn.on("close", () => {
+            this.#friends.delete(friend.id);
+            this.notifyFriendChanges(-1);
+        });
+    }
+
+    /**
+     * Handler for friend messages
+     * @param message The friend message to process
+     */
+    private handleFriendMessage(message: IFriendMessage): void {
+        const numFriendsBefore = this.#friends.size;
+        // List of ids used for cleaning up friends list after updates
+        const newListIds: string[] = [];
+
+        for (const messageFriend of message.friends) {
+            const existingFriend = this.#friends.get(messageFriend.id);
+            const existingFriendPeerJsConnection = existingFriend?.connection;
+
+            // Message will contain the current user id
+            if (messageFriend.id !== this.id) {
+                const newFriendInstance = new Friend(messageFriend.id, messageFriend.displayName, existingFriendPeerJsConnection);
+                this.#friends.set(messageFriend.id, newFriendInstance);
+                newListIds.push(messageFriend.id);
+            }
+        }
+
+        // We've set the new friends, but we need to clean up any friends that weren't in the message
+        this.#friends.forEach(friend => {
+            if (newListIds.indexOf(friend.id) < 0) {
+                // Friend entry wasn't in the list of friends we got from the host
+                this.#friends.delete(friend.id);
+            }
+        });
+
+        const numFriendsAfter = this.#friends.size;
+        if (!this.hasNotifiedPartySize) {
```

Comment:
This `hasNotifiedPartySize` variable feels suuuper weird. I also don't see how it provides anything, because even in the case of the first occurrence, `numFriendsAfter - numFriendsBefore` is still going to be the correct value (`end.length - 0`), right?

Reply:
No because when we connect to the host we're still adding the connection to friends map so when we get the FriendsMessage that's the first time we'll know the number of people in the party.

The reason I decided to add the hasNotifiedPartySize is because I wanted to avoid sending a double notification when someone connects to a host.

Reply:
Hm. It feels weird that sometimes we trigger the notification when the Peer connection is opened ... and other times emit it once we receive a friends list.

Reply:
Yeah I think that work assumed that people would only connect to the host. I'll leave this open for now